### PR TITLE
[SPARK-13249][SQL] Add Filter checking nullability of keys for inner join

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-13249

For inner join, the join key with null in it will not match each other, so we could insert a Filter before inner join (could be pushed down), then we don't need to check nullability of keys while joining.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -23,7 +23,7 @@ import org.apache.spark.sql.catalyst.analysis.{CleanupAliases, EliminateSubQueri
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.Literal.{FalseLiteral, TrueLiteral}
 import org.apache.spark.sql.catalyst.expressions.aggregate._
-import org.apache.spark.sql.catalyst.planning.{ExtractFiltersAndInnerJoins, Unions}
+import org.apache.spark.sql.catalyst.planning._
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: Is it better, `import org.apache.spark.sql.catalyst.planning.{ExtractEquiJoinKeys, ExtractFiltersAndInnerJoins, Unions}`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala [hunk 2]
```diff
@@ -126,11 +126,11 @@ trait HashJoin {
           while (currentHashMatches == null && streamIter.hasNext) {
             currentStreamedRow = streamIter.next()
             val key = joinKeys(currentStreamedRow)
-            if (!key.anyNull) {
-              currentHashMatches = hashedRelation.get(key)
-              if (currentHashMatches != null) {
-                currentMatchPosition = 0
-              }
+            // We do filtering null keys by inserting a Filter before inner join,
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: We need assertion: !key.anyNull?

**Comment by maropu** [comment]
ISTM we can also add `NULL` filters in the one side of left/right-outer joins. Is it wrong?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 2 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: As null are filters by added filter, I think we should be able to skip it.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
@maropu yeah, I think so.

**Comment by viirya** [comment]
ping @davies

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 2 above)
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Understood.

**Comment by viirya** [comment]
cc @davies @marmbrus @liancheng @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -57,6 +57,8 @@ abstract class Optimizer extends RuleExecutor[LogicalPlan] {
       ReplaceDistinctWithAggregate) ::
     Batch("Aggregate", FixedPoint(100),
       RemoveLiteralFromGroupExpressions) ::
+    Batch("Join", Once,
+      AddFilterOfNullForInnerJoin) ::
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We should make this rule idempotent instead of hacking it and making it run only once.  You are loosing the benefits of emergent optimizations with this implementation.

I would directly construct the filter in the left/right child, but only when its not already present in the `constraints` of the child.  This is the whole reason we added the ability to reason about what `constraints` are already present on a subtree.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala [hunk 4]
```diff
@@ -274,7 +274,7 @@ case class BroadcastHashJoin(
          |// generate join key for stream side
          |${keyEv.code}
          |// find matches from HashRelation
-         |$bufferType $matches = $anyNull ? null : ($bufferType)$relationTerm.get(${keyEv.value});
+         |$bufferType $matches = ($bufferType)$relationTerm.get(${keyEv.value});
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: Explain these changes please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala (hunk 2 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: The execution engines correctness should not be predicated on specific optimizations.  That is *really* tight coupling.  If you want to optimize code generation for null handling you should do it based on signals like `nullable` and not on undocumented, implicit contracts between components.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: Even if we were going to make changes like this, they should be done in a different PR and come with some benchmarks to show that you are actually speeding things up.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala]: OK. I will revert this part and try to address this in another PR. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala (hunk 4 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: Will revert this and address that in another PR too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 5]
```diff
@@ -123,6 +123,9 @@ object ExtractEquiJoinKeys extends Logging with PredicateHelper {
         case EqualTo(l, r) =>
           canEvaluate(l, left) && canEvaluate(r, right) ||
             canEvaluate(l, right) && canEvaluate(r, left)
+        case EqualNullSafe(l, r) =>
+          canEvaluate(l, left) && canEvaluate(r, right) ||
+            canEvaluate(l, right) && canEvaluate(r, left)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Since `EqualNullSafe` is added into left/right keys in `ExtractEquiJoinKeys`, we should also exclude it from the returned `condition`.

**Comment by viirya** [comment]
@marmbrus I've addressed the comments. Please see if this change is appropriate. Thanks!

**Comment by viirya** [comment]
ping @marmbrus @davies @rxin

**Comment by viirya** [comment]
@marmbrus @davies @rxin any comments for this? Thanks!

**Comment by viirya** [comment]
@liancheng Can you review this too? I think I've addressed previous comments. Thanks!

**Comment by viirya** [comment]
ping @marmbrus @rxin @davies @liancheng Is this ready to go? Or you have other comments? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 6]
```diff
@@ -57,6 +57,8 @@ abstract class Optimizer extends RuleExecutor[LogicalPlan] {
       ReplaceDistinctWithAggregate) ::
     Batch("Aggregate", FixedPoint(100),
       RemoveLiteralFromGroupExpressions) ::
+    Batch("Join", FixedPoint(100),
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We may have more InnerJoin from OuterJoinElimination, should we move this rule after that?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: If we can make this rule idempotent, we don't need to put this as separate group.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -144,6 +146,56 @@ object EliminateSerialization extends Rule[LogicalPlan] {
 }
 
 /**
+ * Add Filter to left and right of an inner Join to filter out rows with null keys.
+ * So we may not need to check nullability of keys while joining. Besides, by filtering
+ * out keys with null, we can also reduce data size in Join.
+ */
+object AddFilterOfNullForInnerJoin extends Rule[LogicalPlan] with PredicateHelper {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We could also do this for semi join and outer join, should we call it AddFilterOfNullForEquiJoin ?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Not commenting at the name at all (haven't looked at the rest of the pr), but AddFilterOfNullForEquiJoin can be shortened slightly to AddNullFilterForEquiJoin

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Renamed. Will do part of semi join and outer join in separate PR once this getting merged.

**Comment by viirya** [comment]
Comments addressed, please check if the change is good to merge now. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 8]
```diff
@@ -146,6 +147,56 @@ object EliminateSerialization extends Rule[LogicalPlan] {
...
 
 /**
+ * Add Filter to left and right of a EquiJoin to filter out rows with null keys.
+ * So we may not need to check nullability of keys while joining. Besides, by filtering
+ * out keys with null, we can also reduce data size in Join.
+ */
+object AddNullFilterForEquiJoin extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case ExtractEquiJoinKeys(Inner, leftKeys, rightKeys, condition, left, right) =>
+      val leftConditions = leftKeys.distinct.map { l =>
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We should only add predicate if the key is nullable and there is no IsNotNull constraints on the key.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 9]
```diff
@@ -146,6 +147,56 @@ object EliminateSerialization extends Rule[LogicalPlan] {
...
+
+      val rightConditions = rightKeys.distinct.map { r =>
+        IsNotNull(r)
+      }.reduceLeft(And)
+
+      val keysConditions = ExpressionSet(leftKeys.zip(rightKeys).map { lr =>
+        EqualTo(lr._2, lr._1)
+      }).reduceLeft(And)
+
+      val finalConditions = if (condition.isDefined) {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: It's better to use the original condition

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 10]
```diff
@@ -146,6 +147,56 @@ object EliminateSerialization extends Rule[LogicalPlan] {
...
+        Filter(leftConditions, left)
+      }
+
+      val newRight = if (rightHasAllNotNullPredicate) {
+        right
+      } else {
+        Filter(rightConditions, right)
+      }
+
+      Join(newLeft, newRight, Inner, Some(finalConditions))
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: If no filter added, we should return the original join.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 11]
```diff
@@ -146,6 +147,56 @@ object EliminateSerialization extends Rule[LogicalPlan] {
...
+        keysConditions
+      }
+
+      val leftHasAllNotNullPredicate = left.constraints.nonEmpty &&
+        left.constraints.filter(_.isInstanceOf[IsNotNull])
+          .forall(expr => leftConditions.references.intersect(expr.references).nonEmpty)
+
+      val rightHasAllNotNullPredicate = right.constraints.nonEmpty &&
+        right.constraints.filter(_.isInstanceOf[IsNotNull])
+          .forall(expr => rightConditions.references.intersect(expr.references).nonEmpty)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This seems complicated and definitely needs more comments.  Why are we not just adding IsNotNull to the left/right when it doesn't already exist in the conditions?

**Comment by davies** [comment]
I did not realized that we had merged https://github.com/apache/spark/pull/11372/files, do we still need this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 11 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Because we want to eliminate reading nulls in advance of join operator. If we simply add IsNotNull to the left/right to the conditions, the nulls are still read during performing join.

**Comment by viirya** [comment]
@davies yea, looks like it is doing the same. Let me close this now. Thanks for reviewing this anyway!
